Replace result callback with Hooks API (#67)
red-perfume/red-perfume: README.md

### API Example

> You can point to files or pass strings in directly. Tasks are sequential, the output of one can feed into the input of the next. You can output to file or use callback hooks provided (more in the next example).

Author: ```suggestion
You can point to files or pass strings in directly. Tasks are sequential, the output of one can feed into the input of the next. You can output to file or use lifecycle callback hooks (documented in next section).
```

### API Documentation

> Key                      | Type     | Allowed          | Default         | Description
> :--                      | :--      | :--              | :--             | :--
> `verbose`                | Boolean  | `true`, `false`  | `true`          | If true, consoles out helpful warnings and errors using `customLogger` or `console.error`.
> `customLogger`           | Function | Any function     | `console.error` | You can pass in your own custom function to log errors/warnings to. When called the function will receive a `message` string for the first argument and sometimes an `error` object for the second argument. This can be useful in scenarios like adding in custom wrappers or colors in a command line/terminal. This function may be called multiple times before all tasks complete. Only called if `verbose` is true.
> `tasks`                  | Array    | Array of objects | `undefined`     | An array of task objects. Each represents the settings for an atomization task to be performed.

Author: ```suggestion
Key             | Type     | Allowed          | Default         | Description
:--             | :--      | :--              | :--             | :--
`verbose`       | Boolean  | `true`, `false`  | `true`          | If true, consoles out helpful warnings and errors using `customLogger` or `console.error`.
`customLogger`  | Function | Any function     | `console.error` | You can pass in your own custom function to log errors/warnings to. When called the function will receive a `message` string for the first argument and sometimes an `error` object for the second argument. This can be useful in scenarios like adding in custom wrappers or colors in a command line/terminal. This function may be called multiple times before all tasks complete. Only called if `verbose` is true.
`tasks`         | Array    | Array of objects | `undefined`     | An array of task objects. Each represents the settings for an atomization task to be performed.
```

  * `afterValidation` - Right after the options are validated, they will be in this state for the rest of all the hooks
  * `beforeTasks` - Right before we start processing the tasks array
  * `afterTasks` - After the last task as been processed, should be the final hook called. Nothing else happens after this.
* Task hooks
  * `beforeTask` - Ran right before a task starts.
  * `afterTask` - Ran right after a task finishes.
* Styles/Markup hooks:
  * `beforeRead` - Right before we get the string of text from files and/or `data`.
  * `afterRead` - Right after we get the string of text from files and/or `data`. Also right before we process/atomize the string.
  * `afterProcessed` - Right after the string has been atomized. Right before we output it to file if `out` is provided.
* Scripts hooks
  * `beforeOutput` - Right before we write the JSON to disk if `out` is provided.
* Styles/Markup/Scripts hook
  * `afterOutput` - Right after the file has been written to disk if `out` is provided. This is always called and the last thing to happen in a subTask

**Hook argument definitions:**

> * `options` - The options object the user originall passed in (`beforeValidation`) or a modifed version with all API defaults in place (any point `afterValidation`)

Author: ```suggestion
* `options` - The options object the user originally passed in (`beforeValidation`) or a modifed version with all API defaults in place (any point from `afterValidation` and on)
```

Key       | Type     | Default     | Description
:--       | :--      | :--         | :--

> `in`      | Array    | `undefined` | An array of strings to valid paths for CSS files. All files will remain untouched. A new atomized string is produced for `out`/hooks.

Author: ```suggestion
`in`      | Array    | `undefined` | An array of strings to valid paths for CSS files. All files will remain untouched. A new atomized string is produced for `out` and/or hooks.
```

> `data`    | String   | `undefined` | A string of CSS to be atomized. Files are provived via `in` are concatenated with `data` at the end, then atomized and sent to `out`/hooks.

Author: ```suggestion
`data`    | String   | `undefined` | A string of CSS to be atomized. Files provived via `in` are concatenated with `data` at the end, then atomized and sent to `out` and/or hooks.
```

> `out`     | String   | `undefined` | A string file path output. If file exists it will be overwritten with the atomized styles from `in` and/or `data`

Author: ```suggestion
`out`     | String   | `undefined` | A string file path output. If file exists it will be overwritten with the atomized styles from `in` and/or `data`
`hooks`   | Object   | `{}`        | Lifecycle callback hooks (documented in next section)
```

**Markup Task API:**

Key       | Type     | Default     | Description
:--       | :--      | :--         | :--
`in`      | String   | `undefined` | Path to an HTML file to be processed.
`data`    | String   | `undefined` | A string of markup to be processed. This is appended to the end of the `in` file contents if both are provided.

> `out`     | String   | `undefined` | Path where the modified version of the `in` file will be stored. If file already exists, it will be overwritten.

Author: ```suggestion
`out`     | String   | `undefined` | Path where the modified version of the `in` file will be stored. If file already exists, it will be overwritten.
`hooks`   | Object   | `{}`        | Lifecycle callback hooks (documented in next section)
```

**Scripts Task API:**

Key       | Type     | Default     | Description
:--       | :--      | :--         | :--

> `out`     | String   | `undefined` | Path where a JSON object will be stored. The object contains keys (selectors) and values (array of strings of atomized class names). If file already exists, it will be overwritten.

Author: ```suggestion
`out`     | String   | `undefined` | Path where a JSON object (`classMap`) will be stored. The object contains keys (selectors) and values (array of strings of atomized class names). If file already exists, it will be overwritten. Output subject to change.
```

> #### Life Cycle Callback hooks example

Author: ```suggestion
#### Lifecycle Callback Hooks Example
```
